### PR TITLE
Allow cable lift to start after block brake

### PR DESCRIPTION
**Pull request by X123M3-256**
This PR allows the giga coaster cable lift, which normally has to start directly after the station, to begin from a block brake instead. 

It also removes the requirement for the layout to be a complete circuit. Of course, the giga coaster has that requirement anyway, but if you use cheats to set the operating mode to one that will allow it, the cable lift will still work.

The only reason this requirement existed is that the code to initialise the cable lift would scan the entire track layout backward from the station looking for the lift piece, and required a complete circuit to ensure it would be found. This isn't necessary, because the location of the cable lift piece is stored in ride->CableLiftLoc. The new code is also shorter.

Although it's not particularly useful to be able to put a cable lift on a shuttle coaster, it does allow it to work properly with the ghost testing mode. Previously, the cable lifts could only be initialised once the circuit had been completed, so if you tried to test it before then, the cable lift would not work properly. Now it works as it should.

### CODE REVIEW

**Comment by 733737** [comment]
should be able to close https://github.com/OpenRCT2/OpenRCT2/issues/9541 and https://github.com/OpenRCT2/OpenRCT2/issues/10044

**Comment by karst** [comment]
This will also close #15330

**Comment by karst** [comment]
Out of curiousity will this also allow #16620 or would more rework be necessary before this could be implemented?

**Comment by X123M3-256** [comment]
> Out of curiousity will this also allow #16620 or would more rework be necessary before this could be implemented?

No, this doesn't allow that. This relies on the ride->CableLiftLoc field which would probably have to be removed if you wanted to allow more than one cable lift on a ride. The code assumes there is only one in so many places, I don't think it's worth trying to change it given I can't think of a single real world example with multiple cable lifts or launches on one ride.

**Comment by karst** [comment]
I will make sure to give this a test when I am home later today.

**Comment by karst** [comment]
image
When starting a test, while the top of the lift is occupied the pulley still goes down to the block, rather than be connected to the train. The train on top will then roll back and crash into the vehicle below it.
image

Another issue with the block brakes is that it opens the blockzone as soon as the pulley leaves the block, rather than the full train, this allows two trains to occupy one block zone at the same time. It will be rare occassions that something could happen, but it can cause an unintended crash which won't be fun for the players.

Behaviour:
The vehicle stops at the block, even when the pulley is already there, I think this behaviour is perfect for this usecase. 
This also prevents the train from overshooting which great.
If a train hits the blocks too hard, usually it would overshoot it, and this isn't the case now, even when going in it at 150kmh it stops it at the block, I don't know if this is intentional. I do not mind this though as if it would the ride would be a broken ride anyhow.

Over all this feature allows a lot of fun new builds and definitely is a worthy addition to the game.

So in conclusion the following needs fixing:
- Pulley needs to spawn with/stop at train at the tophat if one is located there.
- Blockbrakes need to unlock the blockzone once the entire train is out.

Testpark: [Blockzone cable lift.zip](https://github.com/OpenRCT2/OpenRCT2/files/12501034/Blockzone.cable.lift.zip)

**Review by karst — CHANGES_REQUESTED** [review]
- Pulley needs to spawn with/stop at train at the tophat if one is located there.
- Blockbrakes need to unlock the blockzone once the entire train is out.

The rest is approved by me.

**Comment by karst** [comment]
This still requires a review

**Comment by SpartanFrederic104** [comment]
Again, still needs a review. Please keep this open

**Comment by tupaschoal** [comment]
Before requiring a review, this needs some merge conflicts and issues pointed out by @karst to be tackled. Can you tackle them @X123M3-256 ?

**Comment by karst** [comment]
I don't think X7 will work on any projects anymore, though I am unsure on how he's feeling right now. I think someone else needs to take this off his hands.

File: src/openrct2/ride/CableLift.cpp [hunk 1]
```diff
@@ -325,7 +323,8 @@ bool Vehicle::CableLiftUpdateTrackMotionBackwards()
             SetTrackDirection(output.begin_direction);
             SetTrackType(output.begin_element->AsTrack()->GetTrackType());
 
-            if (output.begin_element->AsTrack()->GetTrackType() == TrackElemType::EndStation)
+            if (output.begin_element->AsTrack()->GetTrackType() == TrackElemType::EndStation
+                || output.begin_element->AsTrack()->GetTrackType() == TrackElemType::BlockBrakes)
```
Comment (Gymnasiast) [inline src/openrct2/ride/CableLift.cpp:327]: Needs to account for diagonal block brakes.

File: src/openrct2/ride/Ride.cpp [hunk 2]
```diff
@@ -3736,97 +3736,42 @@ void Ride::MoveTrainsToBlockBrakes(const CoordsXYZ& firstBlockPosition, TrackEle
...
                 break;
-        }
-        if (isApplying)
-        {
-            auto tmpLoc = CoordsXYZ{ it.current, tileElement->GetBaseZ() };
-            auto direction = tileElement->GetDirection();
-            trackType = tileElement->AsTrack()->GetTrackType();
-            GetTrackElementOriginAndApplyChanges({ tmpLoc, direction }, trackType, 0, &tileElement, flags);
+            case TrackElemType::EndStation:
+            case TrackElemType::BlockBrakes:
```
Comment (Gymnasiast) [inline src/openrct2/ride/Ride.cpp:3771]: Idem

**Comment by X123M3-256** [comment]
OK, so while I haven't done any more work on this PR I'd like to note that another bug has been found by a user on Discord.

This code does not remove the cable lift flag from pieces that should not have it, which means, if you build. If you build a ride with a cable lift that starts from the station, spawn the lift either by turning on the ghost simulation mode or completing and testing the ride, but then edit the ride to move the cable lift later in the layout without deleting the piece immediately after the station, then the ride does not work properly.

The train in the station will wait for the cable lift because it looks at the tile immediately after the station to determine this - the catch car will continue all the way from wherever the lift is placed until it hits the station where the train is waiting, and won't stop at the station or block that should start the lift. 

I have checked, and the original code correctly despawns the original lift pieces if the lift is moved elsewhere and doesn't suffer from this issue. This does create a difficulty though. The previous code required the entire circuit to be complete before spawning the lift and would iterate over the entire layout from the station. I thought it was doing this to unnecessarily search for the cable lift piece, whose location is stored when it's placed, however, doing this also allows it to find all pieces that might have the cable lift flag set.

I do, however, think it's useful to be able to spawn the cable lift with an incomplete circuit because this allows the ghost testing mode to work properly with cable lifts. But I'm not sure of a way to do this that fully solves the issue. Is there a way to iterate over all the track pieces belonging to a ride regardless of whether they are connected or not?

**Comment by Gymnasiast** [comment]
> Is there a way to iterate over all the track pieces belonging to a ride regardless of whether they are connected or not?

You can iterate over the map and checks for tile elements that are tracks and belong to the ride in question. That’s the usual way of doing that. It is used in places like S4Importer and S6Importer.

**Comment by X123M3-256** [comment]
> > Is there a way to iterate over all the track pieces belonging to a ride regardless of whether they are connected or not?
> 
> You can iterate over the map and checks for tile elements that are tracks and belong to the ride in question. That’s the usual way of doing that. It is used in places like S4Importer and S6Importer.

Hmm, would this be preferred? I have finally got back to looking at this and right now, what I'm doing is iterating backwards from the cable lift element, through the whole circuit but not returning an error if the circuit is not complete like it normally would, it just stops and only returns an error if the cable lift doesn't start directly after a station or block brake.

This means that it will correctly despawn all cable lift elements if the ride circuit is complete, but not necessarily if the ride is in simulate mode or has been hacked onto an operating mode that does not force a complete circuit. However, you have to take some very specific steps in a specific order to actually make it break - because you need to have a cable lift piece that is both directly in front of a station and is located in front of the actual cable lift, and it will always work when the ride is tested for real, unless it has been hacked, because the giga coaster only allows continuous circuit mode.

I feel like this is still an improvement over not allowing the cable lift in simulate mode at all because it almost always works, but maybe it would be cleaner to just iterate over the entire map and gurante. However, I believe that, while the original code always iterates through the entire circuit, it will not despawn cable lift pieces that are not connected to the track. I have found this useful for hacking, as you can copy and paste cable lift pieces elsewhere with the tile inspector and not have them despawn when you reopen the ride.

On the other hand, I am doing this because I want to add a cable launch piece in future, and for the cable launch piece, it would be useful for it to work properly on shuttle coasters, such as the looping coaster. Perhaps and alternate solution is that I iterate both forwards *and* backwards from the station, thus clearing any pieces that the train can reach, before spawning in the cable lift.

I guess, if you have track merges, this can *still* fail, but if you have track merges, the existing code can also fail, and it's just so unlikely it would happen unless you're trying to do it (in which case, maybe the broken behaviour is what you're after - I saw one guy asking about a way to make a vertical cable lift and you can hack that this way).


I am not going to try to rebase this branch, I am going to overwrite it with new code derived from my latest hydraulic launch branch, as I have been making further changes there. But I will probably try and fix all the outstanding issues on that branch before I do that, because I want to make sure that the code I have works for the cable launch.

**Comment by X123M3-256** [comment]
OK, I have finally got around to updating this branch to resolve the issues raised. These issues turned out to be bugs that are present on develop (and vanilla as well, I believe), but they did make the new feature nearly unusable.

This fixes #21215 by simply avoiding spawning trains on the cable lift at all. In block sectioned mode there always has to be at least one more block than trains, and you can only have one cable lift, so it always possible to simply skip over the cable lift and place the trains in the other blocks. I think that given the way the cable lift works, allowing a train to spawn on the cable lift and not crash would be a much more difficult fix.

The other bug is that if you build a layout where the cable lift was the last block section before the first station, the game would hang when spawning trains and then place the train in a random spot on the layout. This bug appears to be caused by the fact that the train spawning function sets the last block before the station closed, but it does not set all tiles and so the train does not register the block as closed and gets stuck in a loop until the game gives up. Again, I have dealt with this by simply skipping over the cable lift block when spawning trains.

As the title suggests the main change in this PR is that cable lifts are now permitted to start from a block brake as well as a station; this is a relatively simple change and it's implemented the same way as before, though I have made some minor improvements to it from the previous version. Before, when a train would spawn in the block brake, it would have zero velocity and would crawl forward very slowly until it engages the cable lift. Now, the train's speed is set to 4mph if it is less than that, the same as for a regular block brake.

The other major change I have made is to completely rewrite RideInitialiseCableLift. The previous version would loop over the entire circuit from the station to search for the cable lift piece and clear any existing cable lift pieces, which meant the circuit had to be complete before the cable lift could spawn. The new implementation does not require the circuit to be complete - it will fail only if the cable lift does not start from a station or block brake. This permits the cable lift to be spawned while in simulate mode, even if the circuit is not yet complete, which was not previously possible, making it difficult to test rides with a cable lift.

However, as I described above this does have some caveats. Since I didn't get a reply to my previous comment the way I decided to go with the approach that I felt was most consistent with the current behaviour - my implementation will first clear all cable lift pieces that are connected to the station (and thus reachable by the train). It will then spawn in the new cable lift.

This does have the issue, though, that in simulate mode, you can edit the track while the trains are running and without having to respawn them, which means it's possible to build a cable lift, disconnect it from the track, then spawn another one elsewhere in the layout without the original being despawned. If you then connect the track back up again, this can lead to a situation where the simulation will not function correctly. Clicking the simulate button again will fix this, though, as will entering test mode.

This is an issue that already exists with the simulate mode, because requiring the circuit to be complete doesn't prevent you from spawning a cable lift in and then editing the circuit to be disconnected again, but as the simulate mode is currently not very useable with cable lifts, I doubt anyone has run into it yet. In fact, there are other similar issues with simulate mode, for example, linked brakes can be spawned in, turned off by a train and then disconnected from their block brake leaving them permanently off. 

I didn't receive any reply to my previous comment so I decided not to try and implement a fix, I don't think this really needs addressing as you have to take a very specific set of steps for it to happen, if it does happen you just need to click the simulate button again to fix it, and it will never affect a ride that is open or testing (unless you use the tile inspector to edit track while the ride is running, but that can break a lot of stuff).

**Comment by Gymnasiast** [comment]
My two line notes are still open. If you think they’re not applicable, please comment on them.

File: src/openrct2/ride/CableLift.cpp (hunk 1 above)
Comment (X123M3-256) [reply src/openrct2/ride/CableLift.cpp:331]: I didn't check for a diagonal brake because there's no diagonal cable lift hill that could go with it.

It is possible to hack a cable lift hill with turns, so it wouldn't hurt to check, but I think the vehicle update code would also need to change to handle the diagonal case properly.

On the other hand, if this is merged without that check, and then a diagonal cable lift piece is added later, it technically could break stuff, so I think I will need to do some testing to make sure it handles the diagonal case properly.

**Review by karst — APPROVED** [review]
In the playtest everything seems to be working as expected.

**Comment by X123M3-256** [comment]
@Gymnasiast I have checked and making it correctly handle the case of a diagonal block brake is not as straightforward as adding it to the switch, the train fails to stop at the brake and I think it's because the check assumes that the block brake is a single tile piece. I am sure this could be fixed but it would complicate the code for a case that seems to be of little use without a diagonal cable lift track piece.

There is current no way to get a cable lift that starts from a diagonal brake except by messing with the code, because the cable lift will never spawn on a diagonal and the tile inspector doesn't currently offer the ability to create it manually either. Do you still want me to try and make it handle this case?

**Comment by Gymnasiast** [comment]
@X123M3-256 It’s fine to keep it like it is. Could you leave a comment in the code explaining why it only checks regular block brakes?

**Comment by X123M3-256** [comment]
@Gymnasiast I have rebased the PR and added the requested comment but I am confused by some of the changes that you made in your PR here https://github.com/OpenRCT2/OpenRCT2/pull/23459. Specifically, I don't know why you made it call MapGetTrackElementAt a second time [here](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/ride/Ride.cpp#L3677). I also don't know why you removed the fix for the game hanging when the cable lift is the last block before the station. I would have thought you would want to fix both the vehicle spawning bugs.

Currently I've kept this code the way I had it before, but I don't know if I should make it match develop because I figure there must be a reason that you changed it. I have kept the changes you made to the comments.

**Comment by Gymnasiast** [comment]
We intend to review this shortly after the release. I’m currently still recovering from illness and so we’re very short on review capacity - and we also have some problems with getting the current release out.

**Comment by Gymnasiast** [comment]
@ZeeMaji Could you test this one as well.

**Comment by ZeeMaji** [comment]
Sure!

**Comment by ZeeMaji** [comment]
From what i can tell, existing parks work fine. haven't noticed any issues in old parks with cable lifts.

The new behavior of allowing cable lifts after block sections and allowing the cable lift to work in simulation mode seems to be working as i'd expect.

These error messages should probably be updated to account for being able to have a cable lift after a block section. But otherwise this has my approval.
Screenshot at 2025-01-17 01-26-48
Screenshot at 2025-01-17 01-26-41

**Comment by SpartanFrederic104** [comment]
I would preferably make the new error message be this: "Cable lift must start immediately after a block section"

**Comment by X123M3-256** [comment]
> I would preferably make the new error message be this: "Cable lift must start immediately after a block section"

It would make sense, but I wonder if some users wouldn't understand that a station counts as a block section and might interpret that to mean that a cable lift *must* start after a block brake. Also, the cable lift is not permitted to start directly after a chain lift, even though that also counts as a block section.

So, I am thinking I'll change the string to "Cable lift must start immediately after station or block brake"

**Review by Gymnasiast — APPROVED** [review]

File: data/language/en-GB.txt [hunk 3]
```diff
@@ -2169,7 +2169,6 @@ STR_3136    :Warning: This design will be built with an alternative vehicle type
 STR_3137    :Select Nearby Scenery
 STR_3138    :Reset Selection
 STR_3139    :Cable lift unable to work in this operating mode
-STR_3140    :Cable lift hill must start immediately after station
```
Comment (AaronVanGeffen) [inline data/language/en-GB.txt:2172]: It doesn't matter at this point, but any particular reason for removing the existing string, rather than updating it?

Comment (Gymnasiast) [reply data/language/en-GB.txt:2172]: Seeing how long old translations linger (e.g. when the Russian translation was updated recently) I figured it would be preferable to have the correct string show up in English, rather than an inaccurate translated one.
